src/commands/deleteVmAssociatedResources.ts

```diff
@@ -0,0 +1,68 @@
+/*---------------------------------------------------------------------------------------------
+ *  Copyright (c) Microsoft Corporation. All rights reserved.
+ *  Licensed under the MIT License. See License.txt in the project root for license information.
+ *--------------------------------------------------------------------------------------------*/
+
+import { ComputeManagementClient } from "@azure/arm-compute";
+import { NetworkManagementClient, NetworkManagementModels } from "@azure/arm-network";
+import { ResourceManagementClient } from "@azure/arm-resources";
+import { ISubscriptionContext, parseError } from "vscode-azureextensionui";
+import { ext } from "../extensionVariables";
+import { localize } from "../localize";
+import { createComputeClient, createNetworkClient, createResourceClient } from "../utils/azureClients";
+import { getNameFromId } from "../utils/azureUtils";
+import { nonNullValueAndProp } from "../utils/nonNull";
+
+export async function deleteVmAssociatedResources(
+    context: ISubscriptionContext,
+    resourceGroupName: string,
+    networkNames: string[],
+    diskName: string): Promise<void> {
+
+    const networkClient: NetworkManagementClient = await createNetworkClient(context);
+
+    for (const networkName of networkNames) {
+        const networkInterface: NetworkManagementModels.NetworkInterface = await networkClient.networkInterfaces.get(resourceGroupName, networkName);
+        await networkClient.networkInterfaces.deleteMethod(resourceGroupName, networkName);
+
+        if (networkInterface.ipConfigurations) {
+            for (const ipConfigurations of networkInterface.ipConfigurations) {
+                const publicIpName: string = getNameFromId(nonNullValueAndProp(ipConfigurations.publicIPAddress, 'id'));
+                const subnetId: string = nonNullValueAndProp(ipConfigurations.subnet, 'id');
+                // example of subnet id: '/subscriptions/9b5c7ccb-9857-4307-843b-8875e83f65e9/resourceGroups/linux-vm/providers/Microsoft.Network/virtualNetworks/linux-vm/subnets/default'
+                const virtualNetworkName: string = subnetId.split('/')[8];
+                const subnetName: string = getNameFromId(subnetId);
+                const subnet: NetworkManagementModels.Subnet = await networkClient.subnets.get(resourceGroupName, virtualNetworkName, subnetName);
+                const nsgName: string = getNameFromId(nonNullValueAndProp(subnet.networkSecurityGroup, 'id'));
```

Comment:
> the cleanupservice is kind of specific to us, so I'm not sure if most users will run into this.

I agree users won't have the exact cleanupservice we use, but I take this as a sign that this scenario is more complex/risky than anticipated. For example, other people might have their own tools/services that do similar things as cleanupservice

I don't fully understand how your code picked the _wrong_ nsg, but I do think we need to be extra confident that can't happen to our users and/or we add more confirmation prompts (maybe one per resource).

Reply:
Something else to consider is maybe we scope this scenario down to _just_ VMs created from our extension. For example, you could maybe add a unique tag to each resource when you create it so that we know _exactly_ which resources to delete. It means that in your portal scenario, we just wouldn't prompt to delete the extra resources because the tag doesn't exist 🤷‍♂️

Reply:
It's not picking the _wrong_ nsg, it's picking the one that is associated to the subnet.  There is only one nsg per subnet, so it's definitely the one that the VM is using.

The portal is just creating this nsg as well as the one it would normally create and is choosing to link it to that instead.  You can see that the one that was deleted in the screenshot is back.

![image](https://user-images.githubusercontent.com/5290572/103947626-3fa23b80-50ed-11eb-9a80-a2002352d777.png)

I think if we wanted to be safer, I could verify the RG for all resources is the same as the VM before deleting it.  But that's not necessarily going to delete all associated resources since I don't know if there are other ways users can hook it up.

Reply:
> Tested this with a VM made in the portal and it deleted the wrong nsg

> It's not picking the wrong nsg

:P

> I think if we wanted to be safer, I could verify the RG for all resources is the same as the VM before deleting it.

Yes I want the "safer" option always in this PR, but checking the RG still feels like under-the-covers magic that may not be entirely accurate.

Take a look at [this thread](https://feedback.azure.com/forums/216843-virtual-machines/suggestions/8945692-delete-vm-with-all-associated-resources), which has some interesting tidbits.

> In the document Microsoft states, "Also, in a large infrastructure, identifying the resources which were the part of the VM is very difficult and [sic] to delete manually."

This may confirm my suspicion that the portal knows this is an in-demand feature, but just hasn't done it yet because of the complexity.

> Was there previously a feature that let you select/unselect the associated resources when deleting a vm? Maybe this was in a preview I saw or in a dream.. or maybe a feature from deleting a different type of resource.

This may be another way to do it. You could maybe present the user with a multi-select quick pick of resources you think are related to the VM. At least it would be an explicit choice on the user's part to delete resources